Added object cache for Swift ext storage

PVince81 opened the pull request:
This avoid calling the remote API for repeated calls to is_dir,
filetype, etc

I haven't fully tested this yet, but it does seem to be a bit faster.

I'm starting to think we might want to provide a generic cache for remote storages, because the Amazon implementation would also need it. The Dropbox one also has something similar.

@icewind1991

Commit:
Added object cache for Swift ext storage

This avoid calling the remote API for repeated calls to is_dir,
filetype, etc

MorrisJobke commented:
@PVince81 @DeepDiver1975 I guess this is something for 8.1?

PVince81 commented:
Yes. This only needs review/testing but is a bigger change.
Let's set to 8.1

PVince81 commented:
One alternative would be to do it in a generic way for all storages by implementing a stat cache as storage wrapper: https://github.com/owncloud/core/issues/13971

PVince81 commented:
Ok, turns out the storage wrapper will not help for local function calls because the storage only calls itself, not the wrapper.

So this means that this PR is still valid (and also that a similar approach like this PR is needed for other storages)

LukasReschke commented:
Rebase required?

PVince81 commented:
Rebased. Fortunately only a namespace change.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/edb54964-3954-48ae-8d51-56bdbe23f5be) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/edb54964-3954-48ae-8d51-56bdbe23f5be) completed: **3 new issues, 10 updated code elements**

enoch85 commented:
@PVince81 Sorry, not using External Storage anymore. Would have tested otherwise.

PVince81 commented:
@butonic @icewind1991 any chance for a review ?

icewind1991 commented on apps/files_external/lib/swift.php:
```diff
@@ -145,6 +178,7 @@ public function mkdir($path) {
 			$metadataHeaders = DataObject::stockHeaders(array());
 			$allHeaders = $customHeaders + $metadataHeaders;
 			$this->getContainer()->uploadObject($path, '', $allHeaders);
```
`uploadObject()` returns a `DataObject` which you can put in the cache

PVince81 replied:
I prefer to invalidate it. From my experience the cached object might not contain all final attributes which can only be retrieved the next time. (but I might be mixing this up with S3)

icewind1991 commented on apps/files_external/lib/swift.php:
```diff
@@ -381,8 +427,11 @@ public function touch($path, $mtime = null) {
+			$object = $this->fetchObject($path);
+			if ($object->saveMetadata($metadata)) {
+				// invalidate target object to force repopulation on fetch
+				$this->objectCache->remove($path);
```
is it possible to change the meta data of our existing object so we dont need to get the meta data for this file again

icewind1991 commented on apps/files_external/lib/swift.php:
```diff
@@ -405,8 +456,10 @@ public function copy($path1, $path2) {
+				$source = $this->fetchObject($path1);
 				$source->copy($this->bucket . '/' . $path2);
+				// invalidate target object to force repopulation on fetch
+				$this->objectCache->remove($path2);
```
set the object from `$path1` with it's path changed?

icewind1991 commented on lib/private/cache/arraycache.php:
```diff
@@ -0,0 +1,52 @@
+* Implementation of a volatile cache backed by memory (in form of a simple PHP
+* array) for special purposes.
+*/
+class ArrayCache implements \OCP\ICache {
```
https://github.com/owncloud/core/blob/master/lib/private/memcache/arraycache.php ?

PVince81 replied:
@bantu ^ (the new arraycache impl was your idea)

PVince81 commented:
- [ ] use the existing ArrayCache instead
- [ ] Fix @icewind1991's comments

PVince81 commented:
@icewind1991 actually you might be right with the ability to cache objects after upload and co without rechecking, but will need deeper digging into the API.

Let me rebase this first.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/7c1e7e93-8a1c-4050-a49a-af3e07cf355b) was created.

DeepDiver1975 commented:
hmmm .. in case we move over to flysystem we can make use of their caching mechanisms - I wonder if we should invest time into this right now.

DeepDiver1975 commented:
@PVince81 9.0 ???

PVince81 commented:
Ok, I can revive this for 9.0.
I'm not even sure whether flysystem have a cache mechanism anyway...

Xenopathic commented:
It seems that Flysystem performs no caching at all...

DeepDiver1975 commented:
> It seems that Flysystem performs no caching at all...

aha .... http://flysystem.thephpleague.com/caching/

Xenopathic commented:
@DeepDiver1975 Or that :see_no_evil:

PVince81 commented:
Rebased.

Has failing tests:
```
2) Test\Files\Storage\Swift::testRenameDirectory
Failed asserting that false is true.

/var/www/html/owncloud/tests/lib/files/storage/storage.php:468

3) Test\Files\Storage\Swift::testCopyDirectory
Failed asserting that false is true.

/var/www/html/owncloud/tests/lib/files/storage/storage.php:522
```

PVince81 commented:
Rebased and fixed unit tests.
Also now using the CappedCache for the stat cache.

Please review @owncloud/filesystem 
@owncloud/qa can you help regression test this ? All file operations should still work, this makes SWIFT external storage much faster.

PVince81 commented:
also @tjdett @MorrisJobke in case you are interested

icewind1991 commented:
:+1:

PVince81 commented:
Rebased due to the PSR-4 changes. (no conflicts, automatically resolved)

PVince81 commented:
Ran the Swift tests locally and they passed, merging.
